### ISSUE TITLE
Re-express fitted 0.4.0 spike model in PR #164 (main) form; verify Huber loss is preserved (params approximated from ge_fits.png)

### ISSUE THREAD

**Issue by Haddox** [matsengrp/multidms#281]
# Plan: Express a fitted 0.4.0 spike model in the PR-#164 (`main`) form, with parameters approximated from `ge_fits.png`

## Context

We have a fitted **multidms 0.4.0** model of the SARS-CoV-2 spike DMS data
(`matsengrp/SARS-CoV-2_spike_multidms` @ `6c98b7b`; conditions Delta / Omicron_BA1 /
Omicron_BA2, reference **Omicron_BA1**, Sigmoid GE, shift-lasso `4e-5`, `alpha_d=True`,
gamma left at its default of 0). We want to **re-express that fit in the current `main`
(PR #164) functional form and confirm the Huber loss is preserved by the transformation.**

**Working assumption: the fitted `full_models.pkl` is NOT available** (it is gitignored in the
spike repo). The sigmoid GE parameters therefore cannot be read from any committed file. The
plan's strategy is to **roughly approximate the GE parameters from the committed diagnostic
plot `ge_fits.png`**, combine them with the committed per-mutation and per-variant CSVs, and
show that the 0.4.0 form and the `main` form yield the same Huber loss up to a tiny,
independently-checkable residual.

Both formulations were verified against source: `main` in `multidms/jaxmodels.py`; 0.4.0
recovered from git at `79dc61a:multidms/biophysical.py` + `model.py`.

## The transformation

Write `φ_d(v)` = latent phenotype of variant `v` in condition `d`, `σ` = logistic sigmoid.

**0.4.0** (per condition `d`, identity output, gamma=0):
- latent `φ_d(v) = β0 + α_d^add + Xᵥ·(β + Δ_d)`  (`beta_naught + alpha_d + X@(beta + shift_d)`)
- GE `g(z) = θ_scale·σ(z) + θ_bias`  (`theta["ge_scale"]`, `theta["ge_bias"]`)
- prediction (raw, **no WT subtraction in loss**): `ŷ_d(v) = θ_scale·σ(φ_d(v)) + θ_bias`
- loss: `mean_v huber(ŷ_d(v), y_dv)`, δ=1, mean per condition

**`main`** (`jaxmodels.py`):
- latent `φ_d(v) = β0,d + Xᵥ·β_d`  (independent `Latent(β0, β)` per condition, `jaxmodels.py:189`)
- GE `g(z) = σ(z)` (parameter-free expit, `jaxmodels.py:222`)
- prediction (**WT-subtracted, α-scaled**): `ŷ_d(v) = α·(σ(φ_d(v)) − σ(φ_wt,d))` (`jaxmodels.py:306`)
- loss: `functional_score_loss`, δ=1, mean per condition (`jaxmodels.py:387-412`)

**Parameter map (0.4.0 → main):**
| main parameter | from 0.4.0 | note |
|---|---|---|
| `β_d` (per-condition effect vector) | `β + Δ_d` | reference: `Δ=0 ⇒ β_ref = β` |
| `β0,d` (per-condition intercept) | `β0 + α_d^add` | folds the additive latent offset into the intercept |
| `α` (shared scalar scale) | `θ_scale` | 0.4.0's sigmoid *range* becomes main's output scale |
| — (`θ_bias` has no home) | absorbed by `−α·σ(φ_wt,d)` | cancels iff `D_d = 0` (below) |

**Why the loss is preserved (to within a tiny residual).** With `β_d`, `β0,d` set as above,
`φ_d(v)` is identical variant-for-variant. The two Huber residuals then differ only by a
per-condition constant:

```
residual_0.4.0 − residual_main = θ_bias + θ_scale·σ(φ_wt,d)  ≡  D_d
```

`main` has **no free per-condition output constant** (only the structural `−α·σ(φ_wt,d)`), so
exact equality would require `D_d = 0`. 0.4.0's freely-fit `α_d^add` drives each condition's WT
close to raw-predicted-score 0 (`θ_scale·σ(φ_wt,d) + θ_bias ≈ 0`), which is exactly the identity
`main` builds in structurally. `ge_fits.png` corroborates this: the WT vertical lines sit right
at the `g(φ)=0` crossing (see below). So `residual_main ≈ residual_0.4.0`, and the losses
coincide up to `max_d |D_d|` — exact in the limit `D_d → 0`.

## Reading parameters from `ge_fits.png`

The plot shows observed functional score vs. predicted latent φ, with the black curve
`g(φ)=θ_scale·σ(φ)+θ_bias` (one fit per replicate) and vertical lines at each condition's
`φ_wt,d`. Rough reads (to be refined in the notebook by eyeballing several (φ, g) points off
the curve and solving for the two parameters):

- **`θ_bias ≈ −3.4`** — the lower asymptote (φ→−∞); the curve flattens near y=−3.4 in both panels.
- **`θ_scale ≈ 7–8`** — from the midpoint `g(0)=θ_bias+θ_scale/2 ≈ +0.3 ⇒ θ_scale ≈ 7.4`.
  Uncertain: the upper asymptote is off-screen, so this comes from the midpoint, not both plateaus.
- **`φ_wt,d`** — vertical lines: all three conditions cluster at `φ_wt ≈ −0.4` (replicate 2) and
  `≈ −0.7` (replicate 3).
- **Sanity that `D_d ≈ 0`:** the WT lines coincide with the `g(φ)=0` crossing, i.e.
  `θ_scale·σ(φ_wt,d) + θ_bias ≈ 0`.

These reads are approximate (≈1 sig-fig). The equal-loss *demonstration* is robust to this,
because the 0.4.0-form and main-form losses are computed from the **same** approximated
parameters, so their difference is governed purely by `D_d` (also computed from the same reads),
not by how close the reads are to the true fit.

## Worked example: one real spike variant through both frameworks, end-to-end

Take a real variant of the **reference homolog Omicron_BA1**, `T632N A846V` (replicate 2), with
observed functional score **y = −0.375**. Using the reference homolog keeps the example clean:
its wildtype is the all-zero vector, so there is no homolog "bundle" and the shift is zero
(`Δ_BA1 = 0 ⇒ β_d = β`). The per-mutation effects are the committed fitted values; the sigmoid
shape is read from `ge_fits.png`: `θ_scale = 7.4`, and `θ_bias = −2.97` after calibrating the lower
asymptote so the wildtype sits at predicted score 0 (`D_d = 0`). The reference wildtype latent read
from the plot is `φ_wt = −0.40`, which for the reference is exactly `β0,BA1`.

**Step 1 — build the per-condition effect vector, `β_d = β + Δ_d`** (reference: `Δ = 0`):

| mutation | β (shared) | Δ_BA1 (shift) | β_BA1 = β + Δ |
|---|---|---|---|
| T632N | −0.166 | 0 | −0.166 |
| A846V | −0.267 | 0 | −0.267 |

**Step 2 — latent phenotype.** Each pipeline assembles a latent phenotype φ (the position on
the global-epistasis curve) from the mutations the variant carries. They keep the bookkeeping
differently but arrive at the *same* number:

- **0.4.0** works from one effect vector `β` that is *shared across all homologs*, and adds this
  homolog's shift `Δ_d` on top: `φ = β0 + α_d + Σ (β + Δ_d)`. For the reference homolog the shift
  is zero (`Δ_BA1 = 0`) and the additive offset `α_d` is folded away, so the two mutations
  contribute their shared effects directly on top of the global intercept:
  `φ = −0.40 + (−0.166 − 0.267)`.
- **main** works from this homolog's *own* absolute effect vector `β_d` and its own intercept
  `β0,d`: `φ = β0,d + Σ β_d`. For the reference `β_BA1 = β` and `β0,BA1 = −0.40`, giving the same
  `φ = −0.40 + (−0.166 − 0.267)`.

Either way `φ = −0.833`, so `σ(φ) = 0.303`. The homolog's wildtype sits at `φ_wt = −0.40`
(`σ(φ_wt) = 0.401`) — this is the reference point each pipeline measures the variant against in
Step 3.

*(For a non-reference homolog the reference-frame encoding of every variant additionally carries
that homolog's bundle of defining mutations; those enter both φ and φ_wt and cancel in the
WT-subtraction below, so a variant's predicted score always reflects its effect relative to its
own homolog's wildtype. We use the reference here only so the arithmetic is short.)*

**Step 3 — pass the latent phenotype through the global-epistasis function.** Both pipelines
feed the variant's latent phenotype `φ` through the sigmoid global-epistasis function `σ` to
turn it into a predicted functional score; they differ only in how that sigmoid output is scaled
and referenced:

- **0.4.0** applies its two-parameter sigmoid directly to `φ`:
  `ŷ = θ_scale·σ(φ) + θ_bias` — a raw curve output, with the wildtype subtracted only later when
  reporting effects.
- **main** applies the parameter-free sigmoid and subtracts the wildtype's sigmoid output *inside*
  the prediction, then scales by `α`: `ŷ = α·(σ(φ) − σ(φ_wt))`.

| framework | formula | value |
|---|---|---|
| 0.4.0 (raw sigmoid) | `θ_scale·σ(φ) + θ_bias = 7.4·0.303 − 2.97` | **−0.727** |
| main (WT-subtracted) | `α·(σ(φ) − σ(φ_wt)) = 7.4·(0.303 − 0.401)` | **−0.727** |

The two agree exactly: `θ_bias` in the 0.4.0 form is cancelled by main's `−α·σ(φ_wt)` term,
because `D_BA1 = θ_bias + θ_scale·σ(φ_wt) = 0`.

**Step 4 — Huber loss for this variant (δ = 1):**
`residual = ŷ − y = −0.727 − (−0.375) = −0.352`, so `Huber = ½·(0.352)² = 0.062` — the **same**
in both frameworks.

So this variant contributes identically to the Huber loss under 0.4.0 and under main; summing
that identity over every variant is why the per-condition losses match (Check 1). The equality is
exact because `D_d = 0` — the wildtype sits at predicted score 0. If it did not, every variant in
the condition would be shifted by the same `D_d`, and the two losses would differ by that constant,
which `main` (having no free per-condition offset) cannot absorb.

## Deliverable

Two self-contained artifacts in this repo's `main` environment — no 0.4.0 install, no pkl:

1. **A validation notebook** that demonstrates the loss is preserved (below).
2. **A converter script** that emits, as CSVs, the full parameter set needed to *initialize* a
   `main` `jaxmodels.Model` at the exact point in parameter space implied by the 0.4.0 fit
   (below, "Converter script").

### Validation notebook

`experiments/v04-equivalence/check_equivalence_from_image.ipynb`

Inputs (all committed, fetched read-only via `gh api` from the spike repo @ `6c98b7b`, or
copied locally):
- `results/spike_analysis/mutations_df.csv` — per mutation: `beta`, `shift_Delta`,
  `shift_Omicron_BA2`, plus `dataset_name` (replicate) and `scale_coeff_lasso_shift`
  (select the chosen lasso `4e-5`).
- `results/spike_analysis/training_functional_scores.csv` — `func_score`, `aa_substitutions`,
  `condition`, `replicate`.
- `ge_fits.png` — the GE-parameter reads (recorded as literals in a config cell, with the
  eyeballed values above and a short note on how they were read / their uncertainty).

Steps:
1. **Config cell**: record the read `θ_scale`, `θ_bias`, `φ_wt,d` per replicate/condition, then
   enforce the fit's `D_d ≈ 0` constraint by setting `θ_bias := −θ_scale·σ(φ_wt,ref)` (or
   per-condition), so the parameters are internally consistent (see Ex. 3). Report the raw reads
   AND the calibrated values.
2. **Build the variant encoding** with the current `main` API:
   `mdata = multidms.Data(training_functional_scores.csv, reference="Omicron_BA1",
   collapse_identical_variants="mean", alphabet=multidms.AAS_WITHSTOP_WITHGAP)`; per condition
   get `X` and `x_wt` via `jaxmodels.Data.from_multidms(mdata, cond)` (`jaxmodels.py:50-95`).
   Resolve the replicate-label mismatch (image "replicate 2/3" vs CSV `rep-1/rep-2`) by
   inspecting the CSVs.
3. **Assemble mapped parameters**: `β_d = β + Δ_d` from `mutations_df.csv` aligned to `mdata`'s
   mutation order; `β0,BA1 = φ_wt,BA1`; `β0,d = φ_wt,d − x_wt,d·β_d` for non-reference; `α = θ_scale`.
4. **Reconstruct latents** `φ_d(v) = β0,d + X_v·β_d` for every variant.
5. **Compute both Huber losses per condition** (δ=1, per-condition mean):
   - 0.4.0 form: `mean_v huber(θ_scale·σ(φ) + θ_bias, y)`.
   - main form, two ways: (a) direct formula `mean_v huber(α·(σ(φ) − σ(φ_wt,d)), y)`; (b) through
     the real code path — build `model = jaxmodels.Model(φ={c: jaxmodels.Latent(β0=β0_c, β=β_c)},
     α=jnp.array(θ_scale), logθ={c: jnp.array(0.0)}, reference_condition="Omicron_BA1")`
     (`jaxmodels.py:273-287`) and call `jaxmodels.functional_score_loss(model, data_sets, δ=1.0)`.
6. **Report** a per-condition table: `D_d`, 0.4.0 loss, main loss (formula), main loss (code
   path), |gaps|. Assert (a) formula and code-path main losses agree to machine precision, and
   (b) 0.4.0-vs-main loss gap ≤ threshold set by `max_d |D_d|`.
7. Render the four worked examples above as runnable cells.

### Converter script

`experiments/v04-equivalence/convert_040_to_main_params.py`

**Goal:** given the parameters of a fitted 0.4.0 model, emit CSV files that give *all* parameters
needed to initialize a `main` `jaxmodels.Model` at the same point in parameter space — so a
downstream user can warm-start / evaluate `main` exactly where the 0.4.0 fit sits, without a pkl.

The script applies the parameter map above (`β_d = β + Δ_d`; `β0,d = φ_wt,d − x_wt,d·β_d`;
`α = θ_scale`; `logθ_d = 0`, unused by `functional_score_loss`; `θ_bias` calibrated so `D_d = 0`),
runs it on the spike replicate-2 parameter set, and writes three CSVs under
`experiments/v04-equivalence/main_init_params/`:

- `main_params_beta.csv` — long `(condition, mutation, beta)`, one row per mutation per condition
  → `Model.φ[c].β`.
- `main_params_latent.csv` — per condition `beta0`, `logtheta`, `is_reference` (+ `phi_wt`,
  `D_d` diagnostics) → `Model.φ[c].β0`, `Model.logθ[c]`.
- `main_params_global.csv` — key/value scalars & structural choices (`alpha`,
  `reference_condition`, `global_epistasis=Sigmoid`, `output_activation=IdentityOutput`,
  `huber_delta`, calibrated `theta_bias`) → `Model.α`, `Model.reference_condition`, GE choice.

**Verification (built into the script):** read the CSVs back, rebuild a `jaxmodels.Model`, and
confirm (a) its per-condition `functional_score_loss` matches an in-memory model to machine
precision, and (b) predicted-vs-observed correlations reproduce `func_score_corr.png`.

## Rollout

1. Create a new branch off `main` (e.g. `v04-equivalence-check`).
2. File a GitHub issue whose body is this plan (via `bip.issue.file` / `bip.issue.check`).
3. Then implement the notebook above.
4. Add the converter script and commit its output CSVs.

## Key files / references

- `multidms/jaxmodels.py`: `Model`/`Latent` (273-287, 98-124), `predict_score` (289-309),
  `functional_score_loss` (387-412), `Data.from_multidms` (50-95), `Sigmoid` (217-222).
- `multidms/data.py`: `Data` construction, WT-first sort (~441).
- Spike repo @ `6c98b7b`: `results/spike_analysis/{mutations_df.csv, training_functional_scores.csv}`,
  `ge_fits.png`; config in `spike-analysis.ipynb`.
- 0.4.0 forward formula (for the manual 0.4.0-form loss): `git show 79dc61a:multidms/biophysical.py`
  (`additive_model`, `sigmoidal_global_epistasis`).

## Limits & optional exact path

The absolute loss values are approximate (they inherit the ≈1-sig-fig `ge_fits.png` reads); what
the notebook establishes rigorously is that the **transformation preserves the Huber loss up to
`D_d`**, and that `D_d ≈ 0` for this fit. If an exact numerical match to the true fit is later
wanted, the reliable path is a **deterministic re-fit** under the recorded 0.4.0 config, which
regenerates the exact `theta`/`beta_naught`/`alpha_d`; that is out of scope here.

## Verification

- Notebook assertions pass: main-form loss via direct formula == via `functional_score_loss`
  (machine precision); 0.4.0-vs-main per-condition loss gap ≤ threshold from `max_d |D_d|`.
- The per-condition table shows `D_d ≈ 0` after calibration and near-equal 0.4.0/main losses.
- Sanity: `β_ref` from the map equals the committed shared `β` exactly; `ge_fits.png` WT lines
  align with the `g(φ)=0` crossing.
